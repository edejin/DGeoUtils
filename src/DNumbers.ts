import {DEGREE_TO_PI, DOUBLE_PI_IN_DEGREE, PI_IN_DEGREE, PI_TO_DEGREE} from './DPoint';

const delta = 0.001;

export class DNumbers {
  static like(v: number, s: number, d: number = delta): boolean {
    return Math.abs(v - s) < d;
  }

  static likeZero(v: number): boolean {
    return DNumbers.like(v, 0);
  }

  static like2PI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), DOUBLE_PI_IN_DEGREE);
  }

  static likePI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), PI_IN_DEGREE);
  }

  static rad2Deg(v: number): number {
    return v * DEGREE_TO_PI;
  }

  static deg2Rad(v: number): number {
    return v * PI_TO_DEGREE;
  }
}
