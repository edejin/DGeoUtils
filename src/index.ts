export * from './DCircle';
export * from './DLine';
export * from './DNumbers';
export * from './DPoint';
export * from './DPolygon';
export * from './FastSearch';
export * from './TraceMatrix';
