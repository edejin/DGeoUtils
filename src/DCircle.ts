import {DPoint, EARTH_RADIUS_IN_METERS} from './DPoint';
import {DPolygon} from './DPolygon';
import {DNumbers} from './DNumbers';
import {checkFunction} from './utils';

// eslint-disable-next-line padded-blocks
export class DCircle {

  /**
   * @param [center=(0,0)]
   * @param [r=0]
   */
  // eslint-disable-next-line no-useless-constructor,no-empty-function
  constructor(public center: DPoint = DPoint.zero(), public r: number = 0) {}

  toString(): string {
    return `(${this.center.toString()}, ${this.r})`;
  }

  getValue(): {center: DPoint, r: number} {
    return {center: this.center, r: this.r};
  }

  clone(): DCircle {
    return new DCircle(this.center, this.r);
  }

  /**
   * Find intersection points with other circle.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPoints.png)
   * @param c
   */
  findPoints(c: DCircle): DPoint[] | number {
    if (this.equal(c)) {
      return Infinity;
    }
    const {
      center: {x: x0, y: y0},
      r: r0
    } = this;
    const {
      center: {x: x1, y: y1},
      r: r1
    } = c;

    const r02 = r0 * r0;

    const d = this.center.distance(c.center);
    const a = (r02 - r1 * r1 + d * d) / (2 * d);
    const h = Math.sqrt(r02 - a * a);

    const ad = a / d;
    const dy = y1 - y0;
    const dx = x1 - x0;
    const hd = h / d;

    const x2 = x0 + ad * (x1 - x0);
    const y2 = y0 + ad * (y1 - y0);

    const x31 = x2 + hd * dy;
    const y31 = y2 - hd * dx;
    const x32 = x2 - hd * dy;
    const y32 = y2 + hd * dx;

    const res = [];
    if (!isNaN(x31) && !isNaN(y31)) {
      res.push(new DPoint(x31, y31));
    }
    if (!isNaN(x32) && !isNaN(y32) && !(x31 === x32 && y31 === y32)) {
      res.push(new DPoint(x32, y32));
    }
    return res;
  }

  equal({center, r}: DCircle): boolean {
    return this.center.equal(center) && this.r === r;
  }

  /**
   * Transform circle to polygon
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInside.png)
   * @param [pointCount=64]
   */
  findPolygonInside(pointCount: number = 64): DPolygon {
    const preAngle = 2 * Math.PI / pointCount;
    const points: DPoint[] = [];
    for (let i = 0; i < pointCount; i++) {
      const angle = preAngle * i;
      const x = this.r * Math.cos(angle) + this.center.x;
      const y = this.r * Math.sin(angle) + this.center.y;
      points.push(new DPoint(x, y));
    }
    return new DPolygon([...points, points[0]]);
  }

  /**
   * Transform circle to polygon on sphere. It would be different for different latitude.
   *
   * @remarks
   * Center should be Lng/Lat.
   *
   * @remarks
   * Radius should be in meters.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInsideOnSphere.png)
   * @param [pointCount=64]
   */
  findPolygonInsideOnSphere(pointCount: number = 64): DPolygon {
    checkFunction('findPolygonInsideOnSphere')
      .checkArgument('center')
      .shouldBeDegree(this.center);
    const res = new DPolygon();
    for (let i = 0; i < pointCount; i++) {
      res.push(this.sphereOffset(2 * Math.PI * i / pointCount));
    }
    return res.close();
  }

  private sphereOffset(bearing: number, earthRadius = EARTH_RADIUS_IN_METERS): DPoint {
    const lat1 = DNumbers.deg2Rad(this.center.y);
    const lon1 = DNumbers.deg2Rad(this.center.x);
    const dByR = this.r / earthRadius;
    const lat = Math.asin(Math.sin(lat1) * Math.cos(dByR) +
      Math.cos(lat1) * Math.sin(dByR) * Math.cos(bearing));
    const lon =
      lon1 +
      Math.atan2(
        Math.sin(bearing) * Math.sin(dByR) * Math.cos(lat1),
        Math.cos(dByR) - Math.sin(lat1) * Math.sin(lat)
      );
    return new DPoint(DNumbers.rad2Deg(lon), DNumbers.rad2Deg(lat));
  }
}
