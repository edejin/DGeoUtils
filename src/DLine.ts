import {DPoint} from './DPoint';
import {DCircle} from './DCircle';
import {checkFunction} from './utils';
import {DNumbers} from './DNumbers';

// eslint-disable-next-line padded-blocks
export class DLine {

  /**
   * @param a
   * @param b
   * @param c
   * @param [begin=DPoint.zero()]
   * @param [end=DPoint.zero()]
   */
  constructor(
    public a: number,
    public b: number,
    public c: number,
    public begin: DPoint = DPoint.zero(),
    public end: DPoint = DPoint.zero()
  ) {}

  clone(): DLine {
    return new DLine(this.a, this.b, this.c, this.begin.clone(), this.end.clone());
  }

  findPerpendicular(p: DPoint): DLine {
    checkFunction('findPerpendicular')
      .checkArgument('this.begin')
      .shouldBeMeters(this.begin)
      .checkArgument('this.end')
      .shouldBeMeters(this.end)
      .checkArgument('p')
      .shouldBeMeters(p);
    return new DLine(-this.b, this.a, this.b * p.x - this.a * p.y);
  }

  perpendicularDistance(p: DPoint): number {
    checkFunction('perpendicularDistance')
      .checkArgument('this.begin')
      .shouldBeMeters(this.begin)
      .checkArgument('this.end')
      .shouldBeMeters(this.end)
      .checkArgument('p')
      .shouldBeMeters(p);
    const perpendicularLine = this.findPerpendicular(p);
    const targetPoint = perpendicularLine.findPoint(this);
    return targetPoint!.distance(p);
  }

  /**
   * Find intersection of two lines segments.
   * For intersection of two lines use [[findPoint]]
   * @param l
   * @param [d=0]
   * @param [includeOnly=false]
   */
  intersection(l: DLine, d: number = 0, includeOnly: boolean = false): DPoint | null {
    const p = this.findPoint(l);
    if (p) {
      if (includeOnly) {
        return this.insideRange(p, d) && l.insideRange(p, d) ? p : null;
      }
      return this.inRange(p, d) && l.inRange(p, d) ? p : null;
    }
    return null;
  }

  intersectionWithCircle(circle: DCircle): DPoint | [DPoint, DPoint] | null {
    const {center, r} = circle;
    const per = this.findPerpendicular(center);
    const t = this.findPoint(per)!;
    let distance = t.distance(center);
    if (this.begin.equal(center)) {
      distance = 0;
    }
    if (this.end.equal(center)) {
      distance = 0;
    }
    if (distance < r) {
      const {a, b, c} = this;
      if (this.isParallel) {
        const ct = center.distance(t);
        const move = Math.sqrt(r * r - ct * ct);
        // Mean "|" x = const
        if (this.isParallelY) {
          t.x = this.begin.x;
          const r1 = t.clone().move(0, -move);
          const r2 = t.clone().move(0, move);
          return [r1, r2];
        }
        // Mean "-" y = const
        if (this.isParallelX) {
          t.y = this.begin.y;
          const r1 = t.clone().move(move, 0);
          const r2 = t.clone().move(-move, 0);
          return [r1, r2];
        }
      }
      if (this.begin.like(center)) {
        const p = this.begin.clone();
        return [this.movePoint(p, r), this.movePoint(p, -r)];
      }
      if (this.end.like(center)) {
        const p = this.end.clone();
        return [this.movePoint(p, r), this.movePoint(p, -r)];
      }
      const s = a * a + b * b;
      const d = r * r - c * c / s;
      const mult = Math.sqrt(d / s);
      const r1 = t.clone().move(b * mult, -a * mult);
      const r2 = t.clone().move(-b * mult, a * mult);
      return [r1, r2];
    }
    if (distance === r) {
      return t;
    }
    return null;
  }

  /**
   * Check if point below to line segment
   * @param p
   * @param d
   */
  inRange(p: DPoint, d: number = 0): boolean {
    const {minX, minY, maxX, maxY} = this;
    const {x, y} = p;
    const isInX = x >= minX - d && x <= maxX + d;
    const isInY = y >= minY - d && y <= maxY + d;
    return isInX && isInY;
  }

  /**
   * Check if point below to line segment, but not equal star or end point.
   * @param p
   * @param [d=0]
   */
  insideRange(p: DPoint, d: number = 0): boolean {
    const {begin, end} = this;
    return this.inRange(p, d) && !begin.like(p, 0.00001) && !end.like(p, 0.00001);
  }

  get center(): DPoint {
    return this.begin
      .clone()
      .setIfLessThan(this.end)
      .move(this.end
        .clone()
        .move(this.begin
          .clone()
          .minus())
        .abs()
        .minus()
        .divide(2));
  }

  get minX(): number {
    return Math.min(this.begin.x, this.end.x);
  }

  get minY(): number {
    return Math.min(this.begin.y, this.end.y);
  }

  get maxX(): number {
    return Math.max(this.begin.x, this.end.x);
  }

  get maxY(): number {
    return Math.max(this.begin.y, this.end.y);
  }

  toString(): string {
    return `(${this.a}, ${this.b}, ${this.c})`;
  }

  getValue(): number[] {
    return [this.a, this.b, this.c];
  }

  /**
   * Find `x` from `y` value of point
   * @param p
   */
  x(p: DPoint): DPoint {
    if (this.isParallelY) {
      return new DPoint(-this.c / this.a, p.y);
    }
    if (this.isParallelX) {
      return new DPoint(p.x, -this.c / this.b);
    }
    return new DPoint(-this.b / this.a * p.y - this.c / this.a, p.y);
  }

  /**
   * Find `y` from `x` value of point
   * @param p
   */
  y(p: DPoint): DPoint {
    if (this.isParallelY) {
      return new DPoint(-this.c / this.a, p.y);
    }
    if (this.isParallelX) {
      return new DPoint(p.x, -this.c / this.b);
    }
    return new DPoint(p.x, -this.a / this.b * p.x - this.c / this.b);
  }

  /**
   * Find intersection of two lines.
   * For intersection of two lines segments use [[intersection]]
   * @param l
   */
  findPoint(l: DLine): DPoint | null {
    if (this.isParallelY && l.isParallelY) {
      return null;
    }
    if (this.isParallelX && l.isParallelX) {
      return null;
    }
    if (this.isParallelX && l.isParallelY) {
      return new DPoint(-l.c / l.a, -this.c / this.b);
    }
    if (this.isParallelY && l.isParallelX) {
      return new DPoint(-this.c / this.a, -l.c / l.b);
    }
    if (this.isParallelY) {
      const x = -this.c / this.a;
      return l.y(new DPoint(x));
    }
    if (this.isParallelX) {
      const y = -this.c / this.b;
      return l.x(new DPoint(0, y));
    }
    if (l.isParallelY) {
      const x = -l.c / l.a;
      return this.y(new DPoint(x));
    }
    if (l.isParallelX) {
      const y = -l.c / l.b;
      return this.x(new DPoint(0, y));
    }
    const res = this.y(new DPoint((l.c / l.b - this.c / this.b) / (this.a / this.b - l.a / l.b)));
    if (!isFinite(res.x) && !isFinite(res.y)) {
      return null;
    }
    return res;
  }

  /**
   * Check if line parallel to `x` or `y` axis
   */
  get isParallel(): boolean {
    return this.isParallelX || this.isParallelY;
  }

  get isParallelY(): boolean {
    return Math.abs(this.b) < 0.001;
  }

  get isParallelX(): boolean {
    return Math.abs(this.a) < 0.001;
  }

  /**
   * Get lines segment start and end points as array
   */
  get points(): [DPoint, DPoint] {
    return [this.begin, this.end];
  }

  /**
   * Get line segment direction (from start point to end point)
   */
  getFi(): number {
    checkFunction('getFi')
      .checkArgument('this.begin')
      .shouldBeMeters(this.begin)
      .checkArgument('this.end')
      .shouldBeMeters(this.end);
    const {x, y} = this.end.clone().move(this.begin.clone().minus());
    let v = Math.atan2(y, x) - Math.PI;
    if (v > 0) {
      v = Math.PI - v;
    }
    return (Math.PI - v) % (Math.PI * 2);
  }

  toWKT(): string {
    const {
      begin: {x: x1, y: y1},
      end: {x: x2, y: y2}
    } = this;
    return `LINESTRING (${x1} ${y1}, ${x2} ${y2})`;
  }

  /**
   * Move lines point to left or right
   * @param point
   * @param distance
   */
  movePoint(point: DPoint, distance: number): DPoint;

  /**
   * Move lines point to left or right
   * @param points
   * @param distances
   */
  movePoint(points: DPoint[], distances: number[]): DPoint[];
  movePoint(i: DPoint | DPoint[], k: number | number[]): DPoint[] | DPoint {
    const p: DPoint[] = Array.isArray(i) ? i : [i];
    const d: number[] = Array.isArray(k) ? k : [k];
    const fi = this.findFi(new DLine(1, 0, 0));
    const td = this.x(new DPoint(1, 1)).distance(this.x(new DPoint(2, 2))) / 2;
    const sinCos = new DPoint(Math.sin(fi), Math.cos(fi));
    const dt = sinCos.clone().scale(td);
    const p1T = p[0].clone().move(dt.clone().minus());
    const p2T = p[0].clone().move(dt);
    let res: DPoint[] = [];
    if (DNumbers.like(this.y(p1T).y, p1T.y) || DNumbers.like(this.y(p2T).y, p2T.y)) {
      res = p.map((t: DPoint, index: number) => t.clone().move(sinCos.scale(d[index])));
    } else {
      res = p.map((t: DPoint, index: number) => t.clone()
        .move(sinCos.scale(d[index])
          .setX(({x}) => -x)));
    }
    return res.length === 1 ? res[0] : res;
  }

  /**
   * Find angle between current line and line in argument
   * @param l
   * @param delta
   */
  findFi({a, b}: DLine, delta = 1.0001): number {
    const {a: q, b: w} = this;
    let val = (q * a + w * b) / (Math.sqrt(q * q + w * w) * Math.sqrt(a * a + b * b));
    if (val > 1 && val < delta) {
      val = 1;
    } else if (val < -1 && val > -delta) {
      val = -1;
    }
    return Math.acos(val);
  }

  /**
   * [Cross product](https://en.wikipedia.org/wiki/Cross_product)
   * @param l {DLine}
   */
  vectorProduct({a, b, c}: DLine): DLine {
    const {a: q, b: w, c: e} = this;
    return new DLine(w * c - e * b, e * a - q * c, q * b - w * a);
  }
}
