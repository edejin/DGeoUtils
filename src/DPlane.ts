import {DPoint} from './DPoint';
import {gaussianElimination} from './utils';
import {DPolygon} from './DPolygon';
import {DNumbers} from './DNumbers';
import {DLine} from './DLine';

export class DPlane {
  // eslint-disable-next-line max-params,no-useless-constructor
  constructor(
    public a: number,
    public b: number,
    public c: number,
    public d: number,
    public p1: DPoint = DPoint.zero(),
    public p2: DPoint = DPoint.zero(),
    public p3: DPoint = DPoint.zero()
    // eslint-disable-next-line no-empty-function
  ) {}

  static find(p1: DPoint, p2: DPoint, p3: DPoint): DPlane {
    if (p1.x === p2.x && p2.x === p3.x) {
      return new DPlane(1, 0, 0, -p1.x, p1, p2, p3);
    }
    if (p1.y === p2.y && p2.y === p3.y) {
      return new DPlane(0, 1, 0, -p1.y, p1, p2, p3);
    }
    if (p1.z === p2.z && p2.z === p3.z) {
      return new DPlane(0, 0, 1, -p1.z!, p1, p2, p3);
    }
    const d = 1;
    const [a, b, c] = gaussianElimination([
      [p1.x, p1.y, p1.z!, -d],
      [p2.x, p2.y, p2.z!, -d],
      [p3.x, p3.y, p3.z!, -d]
    ]);
    return new DPlane(a, b, c, d, p1, p2, p3);
  }

  x(p: DPoint): DPoint;
  x(p: DPolygon): DPolygon;
  x(p: DPoint | DPolygon): DPoint | DPolygon {
    if (p instanceof DPoint) {
      const {
        a,
        b,
        c,
        d
      } = this;
      const {
        y,
        z
      } = p;
      p.x = -(b * y + c * z! + d) / a;
      return p;
    }
    return (p as DPolygon).map((t: DPoint) => this.x(t));
  }

  y(p: DPoint): DPoint;
  y(p: DPolygon): DPolygon;
  y(p: DPoint | DPolygon): DPoint | DPolygon {
    if (p instanceof DPoint) {
      const {
        a,
        b,
        c,
        d
      } = this;
      const {
        x,
        z
      } = p;
      p.y = -(a * x + c * z! + d) / b;
      return p;
    }
    return (p as DPolygon).map((t: DPoint) => this.y(t));
  }

  z(p: DPoint): DPoint;
  z(p: DPolygon): DPolygon;
  z(p: DPoint | DPolygon): DPoint | DPolygon {
    if (p instanceof DPoint) {
      const {
        a,
        b,
        c,
        d
      } = this;
      const {
        x,
        y
      } = p;
      p.z = -(a * x + b * y + d) / c;
      return p;
    }
    return (p as DPolygon).map((t: DPoint) => this.z(t));
  }

  clone(): DPlane {
    const {
      a,
      b,
      c,
      d,
      p1,
      p2,
      p3
    } = this;
    return new DPlane(a, b, c, d, p1, p2, p3);
  }

  distance(p: DPoint): number;
  distance(p: DPlane): number;
  distance(p: DPoint | DPlane): number {
    if (p instanceof DPoint) {
      const {
        x,
        y,
        z
      } = p;
      const {
        a,
        b,
        c,
        d
      } = this;
      return Math.abs(a * x + b * y + c * z! + d) / Math.sqrt(a * a + b * b + c * c);
    }
    const {
      a,
      b,
      c,
      d
    } = p as DPlane;
    const {d: r} = this;
    return Math.abs(d - r) / Math.sqrt(a * a + b * b + c * c);
  }

  equal(p: DPlane): boolean {
    const {
      a,
      b,
      c,
      d
    } = p;
    const {
      a: q,
      b: w,
      c: e,
      d: r
    } = this;
    return DNumbers.like(a, q) &&
      DNumbers.like(b, w) &&
      DNumbers.like(c, e) &&
      DNumbers.like(d, r);
  }

  same(p: DPlane): boolean {
    const {
      a,
      b,
      c,
      d
    } = p;
    const {
      a: q,
      b: w,
      c: e,
      d: r
    } = this;
    const t = a / q;
    const y = b / w;
    const u = c / e;
    const i = d / r;
    return DNumbers.like(t, y) &&
      DNumbers.like(t, u) &&
      DNumbers.like(t, c) &&
      DNumbers.like(t, i);
  }

  parallel(p: DPlane): boolean {
    const {
      a,
      b,
      c,
      d
    } = p;
    const {
      a: q,
      b: w,
      c: e,
      d: r
    } = this;
    const t = a / q;
    const y = b / w;
    const u = c / e;
    const i = d / r;
    return DNumbers.like(t, y) &&
      DNumbers.like(t, u) &&
      DNumbers.like(t, c) &&
      !DNumbers.like(t, i);
  }

  findIntersection(p: DPlane): DLine | null {
    if (this.parallel(p)) {
      return null;
    }
    const {a, b, c, d} = this;
    const {a: q, b: w, c: e, d: r} = p;
    const [f, g, h] = gaussianElimination([
      [a, b, c, -d],
      [q, w, e, -r],
      [0, 0, 1, 1]
    ]);
    return new DLine(f, g, h);
  }
}
