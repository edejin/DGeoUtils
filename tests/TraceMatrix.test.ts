/* eslint-disable max-lines,max-statements,max-lines-per-function */
import {DPoint, TraceMatrix, TraceMatrixValues} from '../src';

const data = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
  [0, 1, 1, 0, 0, 1, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 1, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 0, 1, 0, 0, 0],
  [0, 1, 1, 1, 1, 1, 1, 0, 0, 0],
  [0, 1, 1, 1, 1, 1, 0, 0, 0, 0],
  [0, 1, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 1, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];

const data2 = [
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
  [0, 1, 1, 0, 0, 1, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 1, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 0, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 1, 1, 0, 0, 0],
  [0, 1, 1, 0, 1, 1, 0, 0, 0, 0],
  [0, 1, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 1, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];

describe('TraceMatrix', () => {
  const t = new TraceMatrix(
    new DPoint(10, 10),
    ({x, y}: DPoint) => data[y][x] ? TraceMatrixValues.t : TraceMatrixValues.f
  );
  const t2 = new TraceMatrix(
    new DPoint(10, 10),
    ({x, y}: DPoint) => data2[y][x] ? TraceMatrixValues.t : TraceMatrixValues.f
  );

  test('1', () => {
    expect(t.fullMatrixTrace()).toEqual([
      {
        holes: [],
        pPoints: [
          {
            properties: {},
            x: 1,
            y: 2,
            z: undefined
          }, {
            properties: {},
            x: 2,
            y: 2,
            z: undefined
          }, {
            properties: {},
            x: 2,
            y: 4,
            z: undefined
          }, {
            properties: {},
            x: 3,
            y: 5,
            z: undefined
          }, {
            properties: {},
            x: 6,
            y: 1,
            z: undefined
          }, {
            properties: {},
            x: 6,
            y: 5,
            z: undefined
          }, {
            properties: {},
            x: 5,
            y: 6,
            z: undefined
          }, {
            properties: {},
            x: 3,
            y: 6,
            z: undefined
          }, {
            properties: {},
            x: 2,
            y: 8,
            z: undefined
          }, {
            properties: {},
            x: 1,
            y: 8,
            z: undefined
          }, {
            properties: {},
            x: 1,
            y: 2,
            z: undefined
          }
        ],
        properties: {},
        searchStore: {}
      }
    ]);
  });

  test('2', () => {
    expect(t2.fullMatrixTrace().map((l) => l.toWKT())).toEqual([
      'POLYGON ((1 2, 2 2, 2 8, 1 8, 1 2))',
      'POLYGON ((4 3, 6 1, 6 5, 5 6, 4 6, 4 3))'
    ]);
  });
});
