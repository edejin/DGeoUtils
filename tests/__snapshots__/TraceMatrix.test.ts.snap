// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceMatrix 1 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            1,
            1,
          ],
          Array [
            3,
            1,
          ],
          Array [
            3,
            3,
          ],
          Array [
            6,
            3,
          ],
          Array [
            6,
            4,
          ],
          Array [
            7,
            4,
          ],
          Array [
            7,
            9,
          ],
          Array [
            6,
            9,
          ],
          Array [
            6,
            8,
          ],
          Array [
            5,
            8,
          ],
          Array [
            5,
            7,
          ],
          Array [
            4,
            7,
          ],
          Array [
            4,
            5,
          ],
          Array [
            3,
            5,
          ],
          Array [
            3,
            8,
          ],
          Array [
            1,
            8,
          ],
          Array [
            1,
            1,
          ],
        ],
        Array [
          Array [
            5,
            5,
          ],
          Array [
            6,
            5,
          ],
          Array [
            6,
            6,
          ],
          Array [
            5,
            6,
          ],
          Array [
            5,
            5,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;

exports[`TraceMatrix 2 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            1,
            1,
          ],
          Array [
            3,
            1,
          ],
          Array [
            3,
            8,
          ],
          Array [
            1,
            8,
          ],
          Array [
            1,
            1,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            4,
            3,
          ],
          Array [
            6,
            3,
          ],
          Array [
            6,
            4,
          ],
          Array [
            6,
            4,
          ],
          Array [
            7,
            4,
          ],
          Array [
            7,
            9,
          ],
          Array [
            6,
            9,
          ],
          Array [
            6,
            8,
          ],
          Array [
            6,
            8,
          ],
          Array [
            5,
            8,
          ],
          Array [
            5,
            7,
          ],
          Array [
            5,
            7,
          ],
          Array [
            4,
            7,
          ],
          Array [
            4,
            3,
          ],
        ],
        Array [
          Array [
            5,
            5,
          ],
          Array [
            6,
            5,
          ],
          Array [
            6,
            6,
          ],
          Array [
            5,
            6,
          ],
          Array [
            5,
            5,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;

exports[`TraceMatrix 2 negative 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            0,
            0,
          ],
          Array [
            10,
            0,
          ],
          Array [
            10,
            10,
          ],
          Array [
            0,
            10,
          ],
          Array [
            0,
            0,
          ],
        ],
        Array [
          Array [
            1,
            1,
          ],
          Array [
            3,
            1,
          ],
          Array [
            3,
            8,
          ],
          Array [
            1,
            8,
          ],
          Array [
            1,
            1,
          ],
        ],
        Array [
          Array [
            4,
            3,
          ],
          Array [
            6,
            3,
          ],
          Array [
            6,
            4,
          ],
          Array [
            6,
            4,
          ],
          Array [
            7,
            4,
          ],
          Array [
            7,
            9,
          ],
          Array [
            6,
            9,
          ],
          Array [
            6,
            8,
          ],
          Array [
            6,
            8,
          ],
          Array [
            5,
            8,
          ],
          Array [
            5,
            7,
          ],
          Array [
            5,
            7,
          ],
          Array [
            4,
            7,
          ],
          Array [
            4,
            3,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            5,
            5,
          ],
          Array [
            6,
            5,
          ],
          Array [
            6,
            6,
          ],
          Array [
            5,
            6,
          ],
          Array [
            5,
            5,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;

exports[`TraceMatrix 3 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            1,
            1,
          ],
          Array [
            4,
            1,
          ],
          Array [
            4,
            2,
          ],
          Array [
            3,
            2,
          ],
          Array [
            3,
            2,
          ],
          Array [
            3,
            3,
          ],
          Array [
            3,
            3,
          ],
          Array [
            4,
            3,
          ],
          Array [
            4,
            4,
          ],
          Array [
            1,
            4,
          ],
          Array [
            1,
            1,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2,
            7,
          ],
          Array [
            5,
            7,
          ],
          Array [
            5,
            0,
          ],
          Array [
            10,
            0,
          ],
          Array [
            10,
            10,
          ],
          Array [
            2,
            10,
          ],
          Array [
            2,
            7,
          ],
        ],
        Array [
          Array [
            3,
            8,
          ],
          Array [
            5,
            8,
          ],
          Array [
            5,
            9,
          ],
          Array [
            3,
            9,
          ],
          Array [
            3,
            8,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;

exports[`TraceMatrix 3 negative 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            0,
            0,
          ],
          Array [
            5,
            0,
          ],
          Array [
            5,
            7,
          ],
          Array [
            2,
            7,
          ],
          Array [
            2,
            7,
          ],
          Array [
            2,
            10,
          ],
          Array [
            0,
            10,
          ],
          Array [
            0,
            0,
          ],
        ],
        Array [
          Array [
            1,
            1,
          ],
          Array [
            4,
            1,
          ],
          Array [
            4,
            2,
          ],
          Array [
            3,
            2,
          ],
          Array [
            3,
            2,
          ],
          Array [
            3,
            3,
          ],
          Array [
            3,
            3,
          ],
          Array [
            4,
            3,
          ],
          Array [
            4,
            4,
          ],
          Array [
            1,
            4,
          ],
          Array [
            1,
            1,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            3,
            8,
          ],
          Array [
            5,
            8,
          ],
          Array [
            5,
            9,
          ],
          Array [
            3,
            9,
          ],
          Array [
            3,
            8,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;
